source/simulator/core.cc

```diff
@@ -2147,6 +2167,286 @@ namespace aspect
+              // need to update the Stokes preconditioner.
+              rebuild_stokes_matrix = rebuild_stokes_preconditioner = assemble_newton_stokes_matrix = true;
+
+              assemble_stokes_system();
+
+              /**
+               * Eisenstat Walker method for determining the tolerance
+               */
+              if (nonlinear_iteration > 1)
+                {
+                  residual_old = residual;
+                  velocity_residual = system_rhs.block(introspection.block_indices.velocities).l2_norm();
+                  pressure_residual = system_rhs.block(introspection.block_indices.pressure).l2_norm();
+                  residual = std::sqrt(velocity_residual * velocity_residual + pressure_residual * pressure_residual);
+
+                  if (!use_picard)
+                    {
+                      const bool EisenstatWalkerChoiceOne = true;
+                      parameters.linear_stokes_solver_tolerance = compute_Eisenstat_Walker_linear_tolerance(EisenstatWalkerChoiceOne,
+                                                                  parameters.maximum_linear_stokes_solver_tolerance,
+                                                                  parameters.linear_stokes_solver_tolerance,
+                                                                  stokes_residual,
+                                                                  residual,
+                                                                  residual_old);
+
+                      pcout << "   The linear solver tolerance is set to " << parameters.linear_stokes_solver_tolerance << std::endl;
+                    }
+                }
+
+              build_stokes_preconditioner();
+
+              stokes_residual = solve_stokes();
+
+              velocity_residual = system_rhs.block(introspection.block_indices.velocities).l2_norm();
+              pressure_residual = system_rhs.block(introspection.block_indices.pressure).l2_norm();
+              residual = std::sqrt(velocity_residual * velocity_residual + pressure_residual * pressure_residual);
+
+              /**
+               * We may need to do a line search if the solution update doesn't decrease the norm of the rhs enough.
+               * This is done by adding the solution update to the current linearization point and then assembling
+               * the Newton right hand side. If the Newton residual has decreased enough by using the this update,
+               * then we continue, otherwise we reset the current linearization point with the help of the backup and
+               * add each iteration an increasingly smaller solution update until the decreasing residual condition
+               * is met, or the line search iteration limit is reached.
+               */
+              LinearAlgebra::BlockVector backup_linearization_point = current_linearization_point;
+
+              double test_residual = 0;
```

Comment:
Enclose the whole line search code into a set of braces to make sure that all of these variables that you declare here only live until we're done with the line search. It's not necessary for them to live on after we're done with the line search.